core-bundle/src/DependencyInjection/Compiler/RegisterPagesPass.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\DependencyInjection\Compiler;
+
+use Contao\CoreBundle\Routing\Page\CompositionAwareInterface;
+use Contao\CoreBundle\Routing\Page\PageRegistry;
+use Contao\CoreBundle\Routing\Page\PageRouteEnhancerInterface;
+use Contao\CoreBundle\Routing\Page\RouteConfig;
+use Contao\FrontendIndex;
+use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\Compiler\PriorityTaggedServiceTrait;
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Registers Contao pages in the registry.
+ */
+class RegisterPagesPass implements CompilerPassInterface
+{
+    use PriorityTaggedServiceTrait;
+
+    private const TAG_NAME = 'contao.page';
+
+    /**
+     * Adds the fragments to the registry.
+     */
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->has(PageRegistry::class)) {
+            return;
+        }
+
+        $this->registerPages($container);
+    }
+
+    /**
+     * @throws InvalidConfigurationException
+     */
+    protected function registerPages(ContainerBuilder $container): void
+    {
+        $registry = $container->findDefinition(PageRegistry::class);
+
+        foreach ($this->findAndSortTaggedServices(self::TAG_NAME, $container) as $reference) {
+            $definition = $container->findDefinition($reference);
+
+            $tags = $definition->getTag(self::TAG_NAME);
+            $definition->clearTag(self::TAG_NAME);
+
+            foreach ($tags as $attributes) {
+                $routeEnhancer = null;
+                $compositionAware = null;
+                $class = $definition->getClass();
+                $type = $this->getPageType($class, $attributes);
+
+                unset($attributes['type']);
+
+                if (is_a($class, PageRouteEnhancerInterface::class, true)) {
+                    $routeEnhancer = $reference;
+                }
+
+                if (is_a($class, CompositionAwareInterface::class, true)) {
+                    $compositionAware = $reference;
+                }
+
+                $config = $this->getRouteConfig($reference, $definition, $attributes);
```

Comment:
I'd love to have a `RouteConfigAwareInterface` as well. Otherwise the route configuration always have to live in the dependency container. If a controller implements the `RouteConfigAwareInterface` the route config should be retrieved from the controller. Otherwise the current implementation can be used as fallback. Wdyt?

Reply:
That's a bit confusing to me. If the `RouteConfigAwareInterface` is meant to be the same as the `FragmentConfigAwareInterface`, the the interface is there to _set the config on the controller_, so the controller knows about the DI configuration. It is **not** meant to return the config. Returning the config does not make sense, because if you want that, you should return a custom route instead.